src/array.c

```diff
@@ -1009,14 +1032,25 @@ JL_DLLEXPORT void jl_array_del_end(jl_array_t *a, size_t dec)
 JL_DLLEXPORT void jl_array_sizehint(jl_array_t *a, size_t sz)
 {
     size_t n = jl_array_nrows(a);
-    if (sz <= n)
-        return;
-    size_t inc = sz - n;
-    jl_array_grow_end(a, inc);
-    a->nrows = n;
+
+    int min = a->offset + a->length;
+    sz = (sz < min) ? min : sz;
+
+    if (sz <= a->maxsize) {
+        size_t dec = a->maxsize - sz;
+        //if we dont save >10% of maxsize then its not worth it to shrink
```

Comment:
I would make it a factor of 2, since that's the increment that we grow by.

Reply:
I disagree, going from 2 GB to 1.2 GB is a large gain. `sizehint!`-ing to a value shorter than the array is likely rare, and if you do, you likely want to regain the memory.

Reply:
A possible rationale for the different behavior when shrinking vs. when growing is that you strictly need to increase the storage size to `push!` a new element, so it makes sense to directly double the size to avoid doing it repeatedly. OTC, to call `pop!`, you don't strictly need to shrink the storage size: you can perfectly call `resize!` after a bunch of `pop!` calls, and it will be more efficient. So it makes sense to optimize for this case and assume the user really wants the exact requested size.

Reply:
A possible rationale for the different behavior when shrinking vs. when growing is that you strictly need to increase the storage size to `push!` a new element, so it makes sense to directly double the size to avoid doing it repeatedly. OTC, to call `pop!`, you don't strictly need to shrink the storage size: you can perfectly call `resize!` after a bunch of `pop!` calls, and it will be more efficient. So it makes sense to optimize for this case and assume the user really wants the exact requested size.

Reply:
I agree; if somebody specifically `sizehint!`s to a smaller size, the assumption that the array will keep growing doesn't really hold any more. We could compromise on 25% maybe?